Docs: Update style guide with explicit return types (#23918)
grafana/grafana: contribute/style-guides/frontend.md

_SASS styles are deprecated. Please migrate to Emotion whenever you need to modify SASS styles._

### Typing

> In general we let Typescript to infer the types so there is no need to type each variable. There are some exceptions to this:

Reviewer: ```suggestion
In general, you should let Typescript infer the types so that there's no need to explicitly define type for each variable. 

There are some exceptions to this:
```

```typescript
// Typescript needs to know type of arrays or objects otherwise it would infer it as array of any 

// bad
const stringArray = [];

// good
const stringArray: string[] = [];
```

> We specify function return types explicitly in new code. This is to aid readability by being able to tell what function returns just by looking at the signature and also prevents errors when function's return type is broader than author expects. 

Reviewer: ```suggestion
Specify function return types explicitly in new code. This improves readability by being able to tell what a function returns just by looking at the signature. It also prevents errors when a function's return type is broader than expected by the author. 
```

I'm not sure what you mean by "broader" here?

> > Note: We do not have linting for this enabled because of lots of old code that needs to be fixed first.

Reviewer: ```suggestion
> Note: We don't have linting for this enabled because of lots of old code that needs to be fixed first.
```

```typescript
// bad
function transform(value?: string) {
  if (!value) {
    return undefined
  }
  return applyTransform(value)
};

// good
function transform(value?: string): TransformedValue | undefined {
  if (!value) {
    return undefined
  }
  return applyTransform(value)
};
```

> 

Reviewer: Remove unnecessary newlines.
